vault/core.go

```diff
@@ -2335,8 +2336,48 @@ func (c *Core) postUnseal(ctx context.Context, ctxCancelFunc context.CancelFunc,
 	// This is intentionally the last block in this function. We want to allow
 	// writes just before allowing client requests, to ensure everything has
 	// been set up properly before any writes can have happened.
-	for _, v := range c.postUnsealFuncs {
-		v()
+	//
+	// Use a small temporary worker pool to run postUnsealFuncs in parallel
+	postUnsealFuncConcurrency := runtime.NumCPU() * 2
+	if v := os.Getenv("VAULT_POSTUNSEAL_FUNC_CONCURRENCY"); v != "" {
+		pv, err := strconv.Atoi(v)
+		if err != nil || pv < 1 {
+			c.logger.Warn("invalid value for VAULT_POSTUNSEAL_FUNC_CURRENCY, must be a positive integer", "error", err, "value", pv)
+		} else {
+			postUnsealFuncConcurrency = pv
+		}
+	}
+	if postUnsealFuncConcurrency <= 1 {
+		// Out of paranoia, keep the old logic for parallism=1
+		for _, v := range c.postUnsealFuncs {
+			v()
+		}
+	} else {
+		workerChans := make([]chan func(), postUnsealFuncConcurrency)
+		var wg sync.WaitGroup
+		for i := 0; i < postUnsealFuncConcurrency; i++ {
+			wc := make(chan func())
+			workerChans[i] = wc
+			go func() {
+				for {
+					v, more := <-wc
+					if more {
+						wg.Done()
+						v()
+					} else {
+						break
+					}
+				}
+			}()
```

Comment:
```suggestion
			workerChans[i] = make(chan func())
			go func(i int) {
				defer wg.Done()
				for v := range workerChans[i] {
					v()
				}
			}(i)
```

I don't think we should call Done until after the func has run, otherwise we're changing the behaviour.  I also think the use of the local var `wc` might race.

Reply:
Oh, didn't know you could range on a chan. This is cleaner.